File: go/test/batman_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/mcu-countdown-sdk"
	"github.com/voxgig-sdk/mcu-countdown-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestBatmanEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Batman(nil)
		if ent == nil {
			t.Fatal("expected non-nil BatmanEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := batmanBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "batman." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set MCUCOUNTDOWN_TEST_BATMAN_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		batmanRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.batman", setup.data)))
		var batmanRef01Data map[string]any
		if len(batmanRef01DataRaw) > 0 {
			batmanRef01Data = core.ToMapAny(batmanRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = batmanRef01Data

		// LOAD
		batmanRef01Ent := client.Batman(nil)
		batmanRef01MatchDt0 := map[string]any{
			"id": batmanRef01Data["id"],
		}
		batmanRef01DataDt0Loaded, err := batmanRef01Ent.Load(batmanRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		batmanRef01DataDt0LoadResult := core.ToMapAny(batmanRef01DataDt0Loaded)
		if batmanRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if batmanRef01DataDt0LoadResult["id"] != batmanRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func batmanBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "batman", "BatmanTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read batman test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse batman test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"batman01", "batman02", "batman03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("MCUCOUNTDOWN_TEST_BATMAN_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"MCUCOUNTDOWN_TEST_BATMAN_ENTID": idmap,
		"MCUCOUNTDOWN_TEST_LIVE":      "FALSE",
		"MCUCOUNTDOWN_TEST_EXPLAIN":   "FALSE",
		"MCUCOUNTDOWN_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["MCUCOUNTDOWN_TEST_BATMAN_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["MCUCOUNTDOWN_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["MCUCOUNTDOWN_APIKEY"],
			},
			extra,
		})
		client = sdk.NewMcuCountdownSDK(core.ToMapAny(mergedOpts))
	}

	live := env["MCUCOUNTDOWN_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["MCUCOUNTDOWN_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/BatmanEntityTest.php
<?php
declare(strict_types=1);

// Batman entity test

require_once __DIR__ . '/../mcucountdown_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class BatmanEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = McuCountdownSDK::test(null, null);
        $ent = $testsdk->Batman(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = batman_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "batman." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set MCUCOUNTDOWN_TEST_BATMAN_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $batman_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.batman")));
        $batman_ref01_data = null;
        if (count($batman_ref01_data_raw) > 0) {
            $batman_ref01_data = Helpers::to_map($batman_ref01_data_raw[0][1]);
        }

        // LOAD
        $batman_ref01_ent = $client->Batman(null);
        $batman_ref01_match_dt0 = [
            "id" => $batman_ref01_data["id"],
        ];
        [$batman_ref01_data_dt0_loaded, $err] = $batman_ref01_ent->load($batman_ref01_match_dt0, null);
        $this->assertNull($err);
        $batman_ref01_data_dt0_load_result = Helpers::to_map($batman_ref01_data_dt0_loaded);
        $this->assertNotNull($batman_ref01_data_dt0_load_result);
        $this->assertEquals($batman_ref01_data_dt0_load_result["id"], $batman_ref01_data["id"]);

    }
}

function batman_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/batman/BatmanTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = McuCountdownSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["batman01", "batman02", "batman03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("MCUCOUNTDOWN_TEST_BATMAN_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "MCUCOUNTDOWN_TEST_BATMAN_ENTID" => $idmap,
        "MCUCOUNTDOWN_TEST_LIVE" => "FALSE",
        "MCUCOUNTDOWN_TEST_EXPLAIN" => "FALSE",
        "MCUCOUNTDOWN_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["MCUCOUNTDOWN_TEST_BATMAN_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["MCUCOUNTDOWN_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["MCUCOUNTDOWN_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new McuCountdownSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["MCUCOUNTDOWN_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["MCUCOUNTDOWN_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
